ethpm/assets/simple-registry/contracts/PackageRegistry.sol

```diff
@@ -0,0 +1,373 @@
+            string memory manifestURI
+        )
+    {
+        Release memory targetRelease = releases[releaseId];
+        Package memory targetPackage = packages[targetRelease.packageId];
+        return (targetPackage.name, targetRelease.version, targetRelease.manifestURI);
+    }
+
+    /// @dev Returns the release id for a given name and version pair if present on registry.
+    /// @param packageName Package name
+    /// @param version Version string(ex: '1.0.0')
+    function getReleaseId(string memory packageName, string memory version)
+        public
+        view
+        onlyIfPackageExists(packageName)
+        onlyIfReleaseExists(packageName, version)
+        returns (bytes32 releaseId)
+    {
+        return generateReleaseId(packageName, version);
+    }
+
+    /// @dev Returns the number of packages stored on the registry
+    function numPackageIds() public view returns (uint totalCount)
+    {
+        return packageCount;
+    }
+
+    /// @dev Returns the number of releases for a given package name on the registry
+    /// @param packageName Package name
+    function numReleaseIds(string memory packageName)
+        public
+        view
+        onlyIfPackageExists(packageName)
+        returns (uint totalCount)
+    {
+        bytes32 packageId = generatePackageId(packageName);
+        Package storage package = packages[packageId];
+        return package.releaseCount;
+    }
+
+    /// @dev Returns a bool indicating whether the given release exists in this registry.
+    /// @param packageName Package Name
+    /// @param version version
+    function releaseExists(string memory packageName, string memory version)
+        public
+        view
+        onlyIfPackageExists(packageName)
+        returns (bool)
+    {
+        bytes32 releaseId = generateReleaseId(packageName, version);
+        Release storage targetRelease = releases[releaseId];
+        return targetRelease.exists;
+    }
+
+    /// @dev Returns a bool indicating whether the given package exists in this registry.
+    /// @param packageName Package Name
+    function packageExists(string memory packageName) public view returns (bool) {
+        bytes32 packageId = generatePackageId(packageName);
+        return packages[packageId].exists;
+    }
+
+    //
+    //  ====================
+    //  |  Hash Functions  |
+    //  ====================
+    // 
+
+    /// @dev Returns name hash for a given package name.
+    /// @param name Package name
+    function generatePackageId(string memory name)
+        public
+        pure
+        returns (bytes32)
+    {
+        return keccak256(abi.encodePacked(name));
+    }
+
+    // @dev Returns release id that *would* be generated for a name and version pair on `release`.
+    // @param packageName Package name
+    // @param version Version string (ex: '1.0.0')
+    function generateReleaseId(
+        string memory packageName,
+        string memory version
+    )
+        public
+        view
```

Comment:
Should this be `pure`?

Reply:
Yes, it should. But the ERC / Interface define this function as `view`. Which looks to be the case b/c in the complex solidity implementation, it generates these values via the `PackageDB` and `ReleaseDB`. So, I left it as `view` here to maintain compatibility with the erc/ `PackageRegistryInterface`.